omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java

```diff
@@ -147,41 +153,98 @@ protected ModelAndView onSubmit(HttpServletRequest request, HttpServletResponse
+				} else if (secretQuestion != null && Context.getUserService().isSecretAnswer(user, secretAnswer)) {
 					
 					StringBuilder randomPassword = new StringBuilder();
-					for (int i = 0; i < 8; i++) {
-						randomPassword.append(String.valueOf((Math.random() * (127 - 48) + 48)));
+					ArrayList<Character> buildRandomPassword = new ArrayList<Character>();
+
+					//Ensure atleast 1 uppercase, 1 lowercase as well as 1 digit in password (Otherwise, exception is thrown)
+					buildRandomPassword.add((char) (65 + (int) Math.abs(Math.random()) % 26));
+					buildRandomPassword.add((char) (97 + (int) Math.abs(Math.random()) % 26));
+					buildRandomPassword.add((char) (48 + (int) Math.abs(Math.random()) % 10));
+					//Build rest of the password
+					for (int i = 3; i < 8; i++) {
+						buildRandomPassword.add((char) (Math.random() * (127 - 48) + 48));
 					}
-					
+					//Shuffle it
+					Collections.shuffle(buildRandomPassword);
+					//Finally, append it to the string
+					for (char characterInPassword : buildRandomPassword) {
+						randomPassword.append(characterInPassword);
+					}
+
 					try {
 						Context.addProxyPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS);
-						//Context.getUserService().changePassword(user, randomPassword.toString());
+						Context.getUserService().changePassword(user, randomPassword.toString());
 					}
 					finally {
 						Context.removeProxyPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS);
 					}
-					
+
 					httpSession.setAttribute("resetPassword", randomPassword);
 					httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.password.reset");
 					Context.authenticate(username, randomPassword.toString());
 					httpSession.setAttribute("loginAttempts", 0);
+
 					return new ModelAndView(new RedirectView(request.getContextPath() + "/options.form#Change Login Info"));
 				} else {
 					httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.answer.invalid");
 					httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.question.fill");
-					//request.setAttribute("secretQuestion", user.getSecretQuestion());
+					request.setAttribute("secretQuestion", secretQuestion);
 				}
 			}
 		}
 		
 		loginAttemptsByIP.put(ipAddress, forgotPasswordAttempts);
+
+		User testUser = Context.getUserService().getUserByUsername(username);
+		String testSecretQuestion;
+		if (testUser != null) {
+			testSecretQuestion = Context.getUserService().getSecretQuestion(testUser);
+		} else {
+			testSecretQuestion = getRandomFakeSecretQuestion(username);
+		}
 		request.setAttribute("uname", username);
+		request.setAttribute("secretQuestion", testSecretQuestion);
+
 		return showForm(request, response, errors);
 	}
 	
+	public String getRandomFakeSecretQuestion(String username) {
+
+		List<String> questions = new ArrayList<String>();
+		
+		questions.add(Context.getMessageSourceService().getMessage("What is your best friend's name?"));
```

Comment:
These need to be message codes that are defined in the messages.properties file with these as the english translations